src/core/syntax.ml

```diff
@@ -66,16 +66,14 @@ and plain_term =
       (** [match t with p1 |-> t1 | ... | pn |-> tn] *)
   | Conditional of term * term * term  (** [if t then t1 else t2] *)
   | Apply of term * term  (** [t1 t2] *)
-  | Promise of operation * guarded_abstraction * abstraction
-      (** [with op (p1 |-> t1) as p2 in t2] *)
+  | Promise of variable option * operation * abstraction * abstraction
```

Comment:
As we have discussed yesterday, let us keep guarded promises (also in the examples), but express them with recursive promises rather than general recursion.